mautic/mautic
# Autowiring for all!

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [N]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [Y]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation-new/pull/89
| Issue(s) addressed                     | Fixes #... 



#### Description:

Symfony can autowire and autoconfigure services for some time. Here is the doc:

https://symfony.com/doc/4.4/service_container.html

Mautic was still continuing the old way where each service must be configured in the config.php file with all its dependencies. It is a bit daunting to keep all the dependencies up to date in the config file. So this PR is making the developers work a bit nicer.

1. New services do not need to be configured in config.php files
2. Old services can drop their parameters and keep the aliases for BC
3. Did you know that commands were autowired since M3? We still configured them in config.php files...

The biggest problem here is with classes that are not services. Like DTO, value objects, events (which is also DTO) or exceptions. We have to exclude those. In `app/config/services.php` there is common configuration of those excluded directories for all bundles including plugins. But each bundle and plugin can define its own excludes in *Bundle/Config/services.php.

#### Next steps

As this PR is big already I do not want to continue the service refactoring and rather start a new PR for each bundle. This PR enables the refactoring.

1. Finish converting all services in config.php to just a list of aliases in services.php bundle by bundle.
2. Refactor all controllers to use DI
3. Set all services as private by default.

#### Steps to test this PR:

I know this is hard to swallow, but this PR is one of those that can break any single feature in the project.

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
4. The container build would fail if there would be a bug in the changed code. So Mautic wouldn't even load.
5. But I was testing with `bin/console debug:autowiring` command to show me that all services are wired correctly. Clear cache in between runs and changes you make.
6. For example I removed service definition for the command `mautic:segments:update` and service it uses to show that it can be done. So you can run that it works.
7. An issue can come in places where we lazy-load services which is in controllers mainly. So clicking through the UI would be a good test.
